Feature/group sites

ekes opened the pull request:
Switches to use group_sites in place of domain_group; and removes dependency on group_permissions switching to controlling access on routes per domain for creating content.

ekes committed:
WIP

ekes committed:
WIP: Working except pre-domain (permissions) / en-dis-able content types.

ekes committed:
Alternative system to enable/disable modules and content types per microsite.

ekes committed:
Fix tests, and related bugs.

Tests now use domains:
 group-0.{base domain}
 group-1.{base domain}
 group-2.{base domain}
 group-3.{base domain}
in place of the previous a1 b1 etc.

ekes committed:
Need group_sites for additional (controller) permission.

ekes committed:
Fix search_api processor.

ekes committed:
First stab at doing something with the admin mode.

Fix admin mode to be on in the control domain at all times.

ekes committed:
Correct redirect at end of microsite creation.

ekes committed:
A slightly imperfect upgrade path (need to work out how to handle install and uninstall modules)

ekes commented:
To test this probably easiest to checkout the [`feature/group_sites`](https://github.com/localgovdrupal/localgov_microsites_project/tree/feature/group_sites) branch of localgovdrupal/localgov_microsites_project which will then [install the appropriate branches](https://github.com/localgovdrupal/localgov_microsites_project/commit/facefcb66565a951ff661403ca9e14217daa1fe5) down via [localgovdrupal/localgov_microsites#feature/group_sites](https://github.com/localgovdrupal/localgov_microsites/tree/feature/group_sites) and it's [composer.json](https://github.com/localgovdrupal/localgov_microsites/commit/64b1a197e767ff4911a2c1e5c54f22ef75c8753c)

More details on upgrade testing on the issue https://github.com/localgovdrupal/localgov_microsites_group/issues/429#issuecomment-1872992368

finnlewis commented:
My steps to test: 

```
composer create-project --stability dev localgovdrupal/localgov_microsites_project:dev-feature/group_sites MY_PROJECT  --no-install; 
cd MY_PROJECT
ddev start
ddev composer install
ddev drush si localgov_microsites -y
```

The install goes fine, but I am having problems with permissions.

1. log in as user 1
2. create controller user with microsites controller role
3. log in as controller
4. create microsite at https://localgov-microsites.ddev.site/admin/microsites/add/microsite
5. note that the default owner is the controller user
6. try to edit the microsite
7. 403 access denied at https://localgov-microsites.ddev.site/group/3/members or https://localgov-microsites.ddev.site/group/3/edit

It is odd, as I can see the buttons to edit, manage members etc... 

![image](https://github.com/localgovdrupal/localgov_microsites_group/assets/326588/7b247b7a-ace8-42ea-872b-bf7dc3285f17)

ekes committed:
Set group_sites admin mode really early, in a hope to catch all access.

https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1896257326

ekes commented:
So my hunch that it was something I'd done right at the end was correct. It was my attempt to make the control site permanently in 'group_sites admin mode'. Seems it was reporting that it is in admin mode in the toolbar, a brief check for access control in the debugger also saw it confirming admin mode, but _somewhere_, I've not found, it was returning an access denied because it wasn't it seems.

This means with your site as it stands above you should be able to go to the domain itself, login, and administer the microsite from there.

To try and fix this I've now pushed a change that from cursory testing seemed to fix it for me https://github.com/localgovdrupal/localgov_microsites_group/pull/438/commits/9c975ba7daa738c1a5142a65224e56118e706819

But if that doesn't work for you - quite possible it's just a punt that _seems_ to work - to temporarily disable this all and test everything else:
   - comment out the line further down in that file https://github.com/localgovdrupal/localgov_microsites_group/blob/9c975ba7daa738c1a5142a65224e56118e706819/src/EventSubscriber/SiteAdminMode.php#L41
   - then the 'Group sites: Admin Mode OFF/ON' button in the admin menu at the top will become toggleable, you need it to be ON.

finnlewis commented:
Thanks @ekes  - the install is working for me now since that commit https://github.com/localgovdrupal/localgov_microsites_group/commit/9c975ba7daa738c1a5142a65224e56118e706819

If anyone else wants to test this branch, this sets it up quickly for me: 

```
composer create-project --stability dev localgovdrupal/localgov_microsites_project:dev-feature/group_sites MY_PROJECT  --no-install; 
cd MY_PROJECT
ddev start
ddev composer install
ddev drush si localgov_microsites -y
ddev drush uli
```

Adnan-cds commented:
I am finally on it.  Hopefully should be able to complete the review before Tuesday's Merge Monday meeting :D

finnlewis commented:
@nccchris it would be great if you could test this against your multiple microsites.

finnlewis commented:
@ekes  mentioned : on https://github.com/localgovdrupal/localgov_microsites_group/issues/429

> The upgrade path isn't quite there yet. It works, but you need to get the group_sites and group_context_domain enabled before switching to the branch, and you need to have the domain_group and domain_permissions modules there long enough to disable them after switching. I also got an error from localgov_search block, but that might be separate.

So Potential steps to test an upgrade: (this needs testing / confirming)

Assuming we have an existing installation on Drupal 10, (the 3.x branch of https://github.com/localgovdrupal/localgov_microsites)

```
ddev drush pm:uninstall localgov_microsites_permissions
ddev composer require drupal/group_sites drupal/group_context_domain 
ddev composer require drupal/geofield_map
ddev drush en group_sites group_context_domain 
composer require "localgovdrupal/localgov_microsites:dev-feature/group_sites as 2.1.0-beta6" -W 
composer require drupal/domain_group drupal/group_permissions
drush updb
drush cr
composer remove drupal/domain_group drupal/domain_permissions
```

nccchris commented:
First of all there's a small typo : 
ddev composer require geofield_map should be ddev composer require drupal/geofield_map 

I'm probably doing something silly, but I hit a couple of problems. 

At composer require "localgovdrupal/localgov_microsites:dev-feature/group_sites as 2.1.0-beta6" -W 
I get a warning: 
```
drupal/group_permissions has modified files: 
M src/Plugin/Validation/Constraint/UniqueReferenceFieldValidator.php 
Discard changes [y,n,v,d,?]?"
```    
'Y' allows me to proceed, but then at drush updb I hit an error : 

```
[error]  The installed version of the /LocalGov Microsites Group/ module is too old to
update. Update to an intermediate version first (last removed version: 9003,
installed version: 9002).
```

ekes commented:
@nccchris I'm making a first guess here, but is it possible that the site you're upgrading hadn't yet fully upgraded / `run updb` localgov_microsites_group before updating the code.

finnlewis commented:
Note: on my local testing, I can now manage the sites on the control site as a controller. 

But I cannot see any content on each microsite, I just get access denied.

I will re-test from scratch and report steps to reproduce.

Adnan-cds commented:
> My steps to test:

Thanks for the test steps Finn.  All works good on a fresh install.  I will now try the upgrade path.

Drupal logs are getting flooded with PHP 8.2 deprecation notices from the autosave_form module.  A [patch](https://www.drupal.org/project/autosave_form/issues/3355495) is available which we may want.

Adnan-cds commented on composer.json:
```diff
@@ -1,51 +1,49 @@
+        "localgovdrupal/localgov_events": "For events content type in microsites"
+    },
+    "extra": {
+        "enable-patching": true,
```
As far as I remember, `enable-patching` is only relevant for root level composer.json files.

ekes replied:
We're doing this all over localgov. This is possibly something that could go into clear up with https://github.com/localgovdrupal/localgov/issues/597 

Can maybe remove it here already, needs checking though, I think there's some unexpected buggyness with patching from dependencies already.

Adnan-cds commented on localgov_microsites_group.install:
```diff
@@ -17,6 +17,12 @@ function localgov_microsites_group_install($is_syncing) {
+  // Group sites should be installed aleady. It provides configuration, so we
+  // can't, but must update its default config.
+  $group_sites = \Drupal::configFactory()->getEditable('group_sites.settings');
+  $group_sites->set('context_provider', '@group_context_domain.group_from_domain_context:group');
```
Not sure if this is relevant.  But this is what I see at `/admin/group/sites/settings`:
![image](https://github.com/localgovdrupal/localgov_microsites_group/assets/50206849/a2fcb42b-15cc-4fa5-ab4d-07e589f241b4)

ekes replied:
Not related to the install line. That was actually missing from the upgrade, so nice to have it highlighted. The dupe plugin is because one is deprecated. Changing the title of that so it now shows

![image](https://github.com/localgovdrupal/localgov_microsites_group/assets/237186/e4f3126d-ef77-4a3c-996f-697f2801ed14)

Adnan-cds commented on modules/localgov_microsites_group_term_ui/src/Controller/GroupTermUiController.php:
```diff
@@ -102,12 +114,30 @@ public function addTerm(GroupInterface $group, string $vid) {
+   *   The access result.
+   */
+  protected function accessCreateGroupTerm(AccountInterface $account, GroupInterface $group, string $plugin_id) {
+    $plugin_manager = \Drupal::service('group_relation_type.manager');
```
Perhaps this can also be initialized in `GroupTermUiController::create()`.

Adnan-cds commented on localgov_microsites_group.services.yml:
```diff
@@ -40,3 +40,41 @@ services:
+    arguments: ['@domain.negotiator', '@group_sites.admin_mode']
+    tags:
+      - { name: event_subscriber }
+
```
It will be good to have some comments here clarifying the domain_group module as the original source of these services.

ekes replied:
228f194

Adnan-cds commented on localgov_microsites_group.routing.yml:
```diff
@@ -15,3 +15,14 @@ localgov_microsites_group.microsite_admin:
     _controller: '\Drupal\localgov_microsites_group\Controller\MicrositeAdminController::redirectToMicrositeAdmin'
   requirements:
     _custom_access: '\Drupal\localgov_microsites_group\Controller\MicrositeAdminController::access'
+
```
It will be good to have some comments here clarifying the domain_group module as the original source of this route.

ekes replied:
Any particular reason? There's nothing funky with the namespacing or so in this particular case? And it does now become just how LocalGov Microsites works (or rather has its admin interface).

Adnan-cds commented on src/ContentTypeHelper.php:
```diff
@@ -0,0 +1,121 @@
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\group\Entity\GroupInterface;
+use Drupal\group\Plugin\Group\Relation\GroupRelationTypeManagerInterface;
+use Drupal\localgov_roles\RolesHelper;
```
Do we need to include `localgov_roles` as a dependency in localgov_microsites_group.info.yml?

ekes replied:
Looks like that was the incorrect RolesHelper being added to the imported classes f714308 fixing that.

Adnan-cds commented on src/ContentTypeHelperInterface.php:
```diff
@@ -5,23 +5,17 @@
-  const UNKNOWN = 'unknown';
-
   /**
    * Module permissions status: enabled, all permissions in place.
```
Is this class still managing permissions?

ekes replied:
Could use different wording, it's access rather than permissions. Maybe 78c8e02 is better?

Adnan-cds commented on src/ContentTypeHelper.php:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  /**
+   * Return array of all 'group_entity_type:entity_bundles' enabled for a
```
I am surprised this two liner wasn't picked up by the code sniffer :)

ekes replied:
1ee0718

Adnan-cds commented on src/ContextProvider/DomainGroupContext.php:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Sets the current group as a context on domains.
+ *
+ * @deprecated
```
If this is deprecated then what is the current recommendation?

ekes replied:
64c0dab

Adnan-cds commented on src/Controller/DomainGroupAddController.php:
```diff
@@ -129,7 +128,18 @@ public function addForm(GroupTypeInterface $group_type) {
-      $form = $this->formBuilder()->getForm(DomainGroupConfigAdd::class, $group, $extra);
+      // Only create a new domain if we have nothing stored.
+      if (!$domain = $store->get("$store_id:domain")) {
+        $values['type'] = $group_type->id();
```
Do we really need this?

ekes replied:
Ah! You mean specifically line `133`, yer that's a c&p isn't it. 180dc1e

Adnan-cds commented on src/DomainGroupResolver.php:
```diff
@@ -0,0 +1,122 @@
+
+    if (!empty($groups)) {
+      foreach ($groups as $group) {
+        if ($group) {
```
Perhaps `if ($group implements GroupInterface) {` would be better.

ekes replied:
Guessing checking instanceof would be consistent with the method. 9cfc24f

Adnan-cds commented on src/GroupDefaultContent.php:
```diff
@@ -72,6 +62,8 @@ public function generate(GroupInterface $group): ?NodeInterface {
       $plugin_id = 'group_node:' . $default->bundle();
       if ($group->getGroupType()->hasPlugin($plugin_id)) {
         $node = $this->replicator->replicateEntity($default);
+        assert($node instanceof Node);
```
`assert($node instanceof NodeInterface);` perhaps?

ekes replied:
Yep. 735c4f7

Adnan-cds commented on tests/src/Functional/ModuleSettingsFormTest.php:
```diff
@@ -124,6 +125,7 @@ public function testGroupManagementAccess() {
 
     // Test member access.
```
Do group members need access to the Group Domain settings page?

ekes replied:
Ah the difference between the two users here is that one is a member with no role but permission to 'use group_sites_admin_mode' permission only, and the other has a 'microsites-admin' role.

Adnan-cds commented on tests/src/Traits/InitializeGroupsTrait.php:
```diff
@@ -60,18 +61,33 @@ public function createMicrositeGroups(array $settings = [], int $count = 5) {
     }
     $this->domainCreateTestDomains($domains, count($domains));
-  }
+/*
+    // Creating domain site settings.
+    foreach ($this->allTestGroups as $group) {
+      $domain_id = 'group_' . $group->id();
+      $config_id = 'domain.config.' . $domain_id . '.system.site';
+      $config = $this->getConfigFactory()->getEditable($config_id);
+      $config->set('name', $group->label());
+      $config->set('slogan', $group->label() . ' Slogan');
+      $config->set('mail', 'group-' . $group->id() . '@user.com');
+      $config->set('page.front', '/group/' . $group->id());
+      $config->set('page.403', '/denied');
+      $config->set('page.404', '/not-found');
+    }
+    $config->save();
+*/
```
Please ⌫ :)

ekes replied:
80aedc9

Adnan-cds reviewed (commented):
> Some code was move directly out of domain_group into localgov_microsites_group

If these can be marked as such, that may help for future reference.

finnlewis commented:
I tested upgrading our example sites from https://demo.microsites.localgovdrupal.org/

I copied the db down locally, installed, then ran the upgrade.

After disabling localgov_search_block, and updaing the domains to match the local test domains, I am seeing 'access denied' on all content on all microsites.

Even logging in as a 'LocalGov Admin' to site 1, creating a new node (event node in my case) gives access denied to both the logged in 'LocalGov Admin' user and anonymous.

@stephen-cox has just replicated the same behaviour on a fresh install of https://github.com/localgovdrupal/localgov_microsites/releases/tag/3.0.0-alpha1, creating a single site, then flipping to this branch and running the updgrade.

@Adnan-cds have you tried your upgrade yet and if so, do you get similar results?

@ekes is aware.... sorry it doesn't just work @ekes !

Adnan-cds commented:
> have you tried your upgrade yet and if so, do you get similar results?

I am at it now.  Hoping to get a clear idea by the end of the day.

Adnan-cds commented:
Sorry Finn, still working on it.  I am having to update our site to Drupal 10 first!  That's taking sometime.

ekes commented:
> I tested upgrading our example sites from https://demo.microsites.localgovdrupal.org/
...
> Even logging in as a 'LocalGov Admin' to site 1, creating a new node (event node in my case) gives access denied to both the logged in 'LocalGov Admin' user and anonymous.
> 
> @stephen-cox has just replicated the same behaviour on a fresh install of https://github.com/localgovdrupal/localgov_microsites/releases/tag/3.0.0-alpha1, creating a single site, then flipping to this branch and running the updgrade.
...
> @ekes is aware.... sorry it doesn't just work @ekes !

Ah! The upgrade has 'Group from URL' ticked and needs changing to the (correct) group from Domain
![image](https://github.com/localgovdrupal/localgov_microsites_group/assets/237186/2e728c39-3066-4530-8635-c9fc57d01d0e)
Which is being done on install as noted by Adnan https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1481937679 We are of course not installing here. The dupe is there for another reason (the second comes from the legacy context supplied by group domain - I think this to remain just in case people are using it - but needs re-labeling to make clear it is deprecated - this is already set so in the code).

Fixes on the way, and running through the issues Adnan raises.

ekes commented:
> After disabling localgov_search_block,

Are the search indexes getting lost. Why are the search indexes getting lost. We're not doing anything to them are we? :confused:

ekes committed:
Clarify deprecated version of the context plugin.

https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1481937679

ekes committed:
Use injected GroupRelationTypeManager. Resolves https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482771572

ekes committed:
Tidy namespaced RolesHelper. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482848051

ekes committed:
Possibly better wording for constants. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482857827

ekes committed:
Coding standards.

ekes committed:
Coding standards.

ekes committed:
Remove extra c&p line. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482881293

ekes committed:
Coding style consistency. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482888319

ekes committed:
Coding standards. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482911188

ekes committed:
Code tidyup. https://github.com/localgovdrupal/localgov_microsites_group/pull/438#discussion_r1482952473

ekes committed:
Fix upgrade. Include (install) config to set group sites context.

https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1941305517

ekes committed:
Prevent config entites getting deleted by updating dependencies.

Just resaving them while both modules domain_group and the upgraded
localgov_microsites_group seems good enough.

ekes commented:
> > After disabling localgov_search_block,
> 
> Are the search indexes getting lost. Why are the search indexes getting lost. We're not doing anything to them are we? 😕

It'll be because they depend on group_domain because of the plugin (that moves into localgov_microsites_group).

Need to update all the dependencies in config and this seems to do it. ad44bea

ekes commented:
OK. I'd say that's ready for review again @stephen-cox @nccchris Oh! And @Adnan-cds I think this might upgrade directly from 2.x you know, it seems I'd forked it from there not 3.x! Tests just done were from 3.x though.

Adnan-cds commented:
I was following Finn's [update instructions given above](https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1921579761).  This is what happened at the `drush updb` stage:
```
------------- ----------- --------------- -----------------------------------
  Module        Update ID   Type            Description
 ------------- ----------- --------------- -----------------------------------
  localgov_mi   9004        hook_update_n   9004 - Migrate microsite domains
  crosites_gr                               to group_context_domain from
  oup                                       domain_group.
  localgov_mi   9005        hook_update_n   9005 - Update configuration
  crosites_gr                               dependencies.
  oup
  localgov_mi   9006        hook_update_n   9006 - Disable domain_group and
  crosites_gr                               group_permissions.
  oup
  localgov_mi   9007        hook_update_n   9007 - Set group sites context as
  crosites_gr                               would be done on install.
  oup
 ------------- ----------- --------------- -----------------------------------


 Do you wish to run the specified pending updates? (yes/no) [yes]:
 >

>  [notice] Update started: localgov_microsites_group_update_9004
>  [notice] Update completed: localgov_microsites_group_update_9004
>  [notice] Update started: localgov_microsites_group_update_9005
>  [notice] Update completed: localgov_microsites_group_update_9005
>  [notice] Update started: localgov_microsites_group_update_9006
>  [error]  The following reasons prevent the modules from being uninstalled: There is content for the entity type: Group permission. <a href="/admin/modules/uninstall/entity/group_permission">Remove group permission entities</a>.
>  [error]  Update failed: localgov_microsites_group_update_9006
 [error]  Update aborted by: localgov_microsites_group_update_9006 
 [error]  Finished performing updates. 

```
This is on a site using localgovdrupal/localgov_microsites:3.0.0-alpha1 with more than one existing microsite.

ekes committed:
Remove group_permisison entities before removing module.

https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1946929456

ekes commented:
> I was following Finn's [update instructions given above](https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1921579761). This is what happened at the `drush updb` stage:
...
> ```
> >  [notice] Update started: localgov_microsites_group_update_9006
> >  [error]  The following reasons prevent the modules from being uninstalled: There is content for the entity type: Group permission. <a href="/admin/modules/uninstall/entity/group_permission">Remove group permission entities</a>.
> ```
> 
> This is on a site using localgovdrupal/localgov_microsites:3.0.0-alpha1 with more than one existing microsite.

Guessing https://github.com/localgovdrupal/localgov_microsites_group/pull/438/commits/700cf1a20d48a9c3e1f8b29b8d30f15cde5bb42d should fix that for you?

stephen-cox commented:
If I do an initial install from https://github.com/localgovdrupal/localgov_microsites_group/pull/439, create a microsite and run through the upgrade instructions, the upgrade works without any errors.

We do need to make the upgrade process as painless as possible.

Adnan-cds commented:
I can confirm that following Finn's [update instructions](https://github.com/localgovdrupal/localgov_microsites_group/pull/438#issuecomment-1921579761) results in a successful update and microsites appear to be okay on the updated instance.

I have noticed just one error message in the Drupal log: `RuntimeException while trying to render item entity:node/NNN:en with view mode search_index for search index Sitewide search: Failed to start the session because headers have already been sent by "/path/to/codebase/vendor/symfony/http-foundation/Response.php" at line 1317. in Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage->start() (line 132 of /path/to/codebase/vendor/symfony/http-foundation/Session/Storage/NativeSessionStorage.php).`  Not sure if [output buffering](https://www.drupal.org/node/3298550), which we had to activate, has anything to do with this.

finnlewis reviewed (approved):
Testing has been going well and we discussed this in our tech group meeting yesterday. The general agreement to merge this and tag a 4.x-alpha release to gain visibility and make further testing easier. Nice work on this so far everyone, thank you!
